### PR TITLE
Adds Fast is better than free protocol for adversarial training

### PR DESCRIPTION
**Pull request by ambrishrawat**
# Description

This pull request adds a new adversarial trainer - [FBF](https://openreview.net/forum?id=BJx040EFvH) (Fast is Better Than Free) to ART. 

Two small bug-fixes in PGD implementation and conftest.py

## Type of change

Please check all relevant options.

- [X] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

I have added 1 unit test for fit and predict of the introduced trainer. 

**Test Configuration**:
- OS MacOS Mojave 10.14.6
- Python version 3.6
- ART version or commit number 1.3.0
- Pytorch

# Checklist

- [ ] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/defences/trainer/adversarial_trainer_FBF.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2018
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```

File: art/defences/trainer/adversarial_trainer_FBF.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+| Paper link: https://openreview.net/forum?id=BJx040EFvH
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF.py:19]: Please add a short description of the content to this module docstring.

File: art/defences/trainer/adversarial_trainer_FBF.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import abc
+from art.defences.trainer.trainer import Trainer
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF.py:24]:
```suggestion
import abc

from art.defences.trainer.trainer import Trainer
```

File: art/defences/trainer/adversarial_trainer_FBF.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+
+class AdversarialTrainerFBF(Trainer, abc.ABC):
+    """
+    This is abstract class for different backend-specific implementations of Fast is better than free protocol
+    for adversarial training
+    | Paper link: https://openreview.net/forum?id=BJx040EFvH
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF.py:31]:
```suggestion
    This is abstract class for different backend-specific implementations of Fast is better than free protocol
    for adversarial training
    | Paper link: https://openreview.net/forum?id=BJx040EFvH
```
```suggestion
    This is abstract class for different backend-specific implementations of Fast is Better than Free protocol
    for adversarial training.
    
    | Paper link: https://openreview.net/forum?id=BJx040EFvH
```

File: examples/adversarial_training_FBF.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+from art.data_generators import KerasDataGenerator
+from art.defences.trainer import AdversarialTrainerFBFPyTorch
+from art.utils import load_cifar10
+
+import torch.nn.functional as F
+from art.classifiers import PyTorchClassifier
+import numpy as np
+import torch
+import torch.nn as nn
+import math
```
Comment (beat-buesser) [inline examples/adversarial_training_FBF.py:15]: Sort imports according to PEP8
```suggestion
import math

import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
from keras.preprocessing.image import ImageDataGenerator

from art.classifiers import PyTorchClassifier
from art.data_generators import KerasDataGenerator
from art.defences.trainer import AdversarialTrainerFBFPyTorch
from art.utils import load_cifar10
```

File: examples/adversarial_training_FBF.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+trainer = AdversarialTrainerFBFPyTorch(classifier, eps=epsilon, use_amp=False)
+
+# Build a Keras image augmentation object and wrap it in ART
+batch_size = 128
+datagen = ImageDataGenerator(
+    horizontal_flip=True
+)
+
+datagen.fit(x_train)
+art_datagen = KerasDataGenerator(
```
Comment (beat-buesser) [inline examples/adversarial_training_FBF.py:181]: Could we use PyTorchDataGenerator instead of KerasDataGenerator?

File: examples/adversarial_training_FBF.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    datagen.flow(x=x_train, y=y_train, batch_size=batch_size, shuffle=True),
+    size=x_train.shape[0],
+    batch_size=batch_size,
+)
+
+# Step 5: fit the trainer
+trainer.fit_generator(art_datagen, nb_epochs=30)
```
Comment (beat-buesser) [inline examples/adversarial_training_FBF.py:188]: Would it be possible to extend this example here with and evasion attack to show the adversarial accuracy of the original and robust models?

File: art/defences/trainer/adversarial_trainer_FBF.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    def __init__(self, classifier, eps=8, **kwargs):
+        """
+        Create an :class:`.AdversarialTrainerFBF` instance.
+
+        :param classifier: Model to train adversarially.
+        :type classifier: :class:`.Classifier`
+        :param eps: Maximum perturbation that the attacker can introduce.
+        :type eps: `float`
+
+        """
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF.py:43]: I think we should add a call to `__init__` of the parent classes.

File: art/defences/trainer/adversarial_trainer_FBF_Pytorch.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2018
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF_Pytorch.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```

File: art/defences/trainer/adversarial_trainer_FBF_Pytorch.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+| Paper link: https://openreview.net/forum?id=BJx040EFvH
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF_Pytorch.py:19]: Could you please add a description of this module to this docstring?

File: art/defences/trainer/adversarial_trainer_FBF_Pytorch.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+import numpy as np
+
+from art.config import ART_NUMPY_DTYPE
+from art.defences.trainer.adversarial_trainer_FBF import AdversarialTrainerFBF
+
+from art.utils import random_sphere
+import torch
+import time
+import torch.nn as nn
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF_Pytorch.py:33]: Sort imports according to PEP8
```suggestion
import logging
import time

import numpy as np
import torch
import torch.nn as nn

from art.config import ART_NUMPY_DTYPE
from art.defences.trainer.adversarial_trainer_FBF import AdversarialTrainerFBF
from art.utils import random_sphere
```

File: art/defences/trainer/adversarial_trainer_FBF_Pytorch.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class AdversarialTrainerFBFPyTorch(AdversarialTrainerFBF):
+    """
+    | Paper link: https://openreview.net/forum?id=BJx040EFvH
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF_Pytorch.py:40]: Could you please add a short description of this class to this docstring?

File: art/defences/trainer/adversarial_trainer_FBF_Pytorch.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+        :param y: Labels for the training set.
+        :type y: `np.ndarray`
+        :param validation_data: Tuple consisting of validation data
+        :type validation_data: `np.ndarray`
+        :param batch_size: Size of batches.
+        :type batch_size: `int`
+        :param nb_epochs: Number of epochs to use for trainings.
+        :type nb_epochs: `int`
+        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function of
+               the target classifier.
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF_Pytorch.py:76]:
```suggestion
        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function of
                                  the target classifier.
```

File: art/defences/trainer/adversarial_trainer_FBF_Pytorch.py [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+        """
+        Train a model adversarially with FBF protocol using a data generator.
+        See class documentation for more information on the exact procedure.
+
+        :param generator: Data generator.
+        :type generator: :class:`.DataGenerator`
+        :param nb_epochs: Number of epochs to use for trainings.
+        :type nb_epochs: `int`
+        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function of
+               the target classifier.
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_FBF_Pytorch.py:139]:
```suggestion
        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function of
                                  the target classifier.
```

File: examples/adversarial_training_FBF.py (hunk 6 above)
Comment (ambrishrawat) [reply examples/adversarial_training_FBF.py]: Good point, I have it changed.

**Comment by beat-buesser** [comment]
@ambrishrawat Thank you very much for implementing a very fast adversarial training protocol!
